stdlib/public/Concurrency/Actor.cpp

```diff
@@ -0,0 +1,931 @@
+    if (!currentActor)
+      break;
+  }
+
+  frame.exit();
+
+  // If we still have an active actor, we should give it up.
+  if (currentActor)
+    currentActor->giveUpThread();
+}
+
+void ProcessInlineJob::process(Job *job, ExecutorRef _executor) {
+  DefaultActorImpl *actor = DefaultActorImpl::fromInlineJob(job);
+
+  // Pull the priority out of the job before we do anything that might
+  // invalidate it.
+  auto targetPriority = job->getPriority();
+
+  // FIXME: force tail call
+  return processDefaultActor(actor, targetPriority);
+}
+
+void ProcessOverrideJob::process(Job *job, ExecutorRef _executor) {
+  auto self = cast<ProcessOverrideJob>(job);
+
+  // Wait until we're ready.
+  bool isOverridden = true;
+  self->Lock.withLockOrWait(self->Queue, [&] {
+    isOverridden = self->IsOverridden;
+    return self->IsReady;
+  });
+
+  // Pull the actor and priority out of the job.
+  auto actor = self->Actor;
+  auto targetPriority = self->getPriority();
+
+  // Destroy the job.
+  delete self;
+
+  // If the job was flagged as overridden, exit immediately without
+  // trying to access the actor.  This is best-effort; we might get
+  // overridden after taking over execution.
+  if (isOverridden)
+    return;
+
+  // FIXME: force tail call
+  return processDefaultActor(actor, targetPriority);
+}
+
+void DefaultActorImpl::enqueue(Job *job) {
+  auto oldState = CurrentState.load(std::memory_order_relaxed);
+
+  ProcessOverrideJob *overrideJob = nullptr;
+
+  while (true) {
+    auto newState = oldState;
+
+    // Put the job at the front of the job list.
+    setNextJobInQueue(job, oldState.FirstJob);
+    newState.FirstJob = JobRef::getUnpreprocessed(job);
+
+    auto oldStatus = oldState.Flags.getStatus();
+    bool wasIdle = oldStatus == Status::Idle;
+
+    // Update the priority.
+    auto oldPriority = oldState.Flags.getMaxPriority();
+    auto newPriority =
+      wasIdle ? job->getPriority()
+              : std::max(oldPriority, job->getPriority());
+    newState.Flags.setMaxPriority(newPriority);
+
+    // If we need an override job, create it (if necessary) and
+    // register it with the queue.
+    bool needsOverride = !wasIdle && newPriority != oldPriority;
+    if (needsOverride) {
+      if (overrideJob) {
+        overrideJob->NextJob = newState.FirstJob;
+      } else {
+        overrideJob = new ProcessOverrideJob(this, newPriority,
+                                             newState.FirstJob);
+      }
+      newState.FirstJob = JobRef::getOverride(overrideJob);
+    }
+
+    // Make sure the status is at least Scheduled.  We'll actually
+    // schedule the job below, if we succeed at this.
+    if (wasIdle) {
+      newState.Flags.setStatus(Status::Scheduled);
+    }
+
+    // Try the compare-exchange, and try again if it fails.
+    if (!CurrentState.compare_exchange_weak(oldState, newState,
+          /*success*/ std::memory_order_release,
+          /*failure*/ std::memory_order_relaxed))
```

Comment:
An enqueuing thread doesn't need to synchronize with the current state of the actor because it never actually looks at any of the memory there besides what it reads from the atomic itself.  It just puts the pointer it finds as the next pointer after the job it's enqueuing.  The sequencing of operations on the atomic is sufficient to order things.

Let me put that in a picture.  The simple enqueue/dequeue scenario looks like this:

```text
Thread #1    ...enqueue....\......................................
                            \ (release)
Actor atomic ................\..................\.................
                                                 \ (acquire)
Thread #2    .....................................\....dequeue....
```

Everything is properly ordered because of the release/acquire and the sequencing of operations on the atomic.

What you're suggesting is that we need to acquire before releasing in the enqueue, so that we have a linear ordering through the second enqueuing thread:

```text
Thread #1    ...enqueue....\.........................................................
                            \ (rel)
Actor atomic ................\.........\.................../.......\.................
                                        \ (acq)           / (rel)   \ (acq)
Thread #2    ............................\....enqueue..../....       \        .......
                                                                      \
Thread #3    ..........................................................\...dequeue...
```

This is clearly well-ordered; you can trace a nice line through both enqueuing threads to the dequeuing thread, and everything interchange on the atomic fully synchronizes.  But that full synchronization isn't actually important because the enqueuing threads never access any of the memory associated with other enqueued jobs, just the atomic object itself and the pointer it contains.  And reads from the atomic object are ordered after all preceding changes to it, not just the last one.  So this simpler diagram is also perfectly well synchronized:

```text
Thread #1    ...enqueue....\.........................................................
                            \ (rel)
Actor atomic ................\............................./.......\.................
                                                          / (rel)   \ (acq)
Thread #2    .................................enqueue..../....       \        .......
                                                                      \
Thread #3    ..........................................................\...dequeue...
```

(Of course there's still a relaxed load that we're not drawing here, and that does have to be ordered on the atomic after the last store, but the compare-exchange takes care of that: if it wasn't ordered on the atomic after the last store, the comparison would fail.)

And there's a very important reason that we can't access any of this indirect memory associated with the queue while enqueuing, which is that that memory is in fact only *temporarily* associated with the queue for the period while it's in the queue, and as soon as some thread dequeues it, it's not safe to access anymore.  Since an enqueue can never guarantee that that isn't happening concurrently, it can never touch any of that memory.  Since we can't touch it anyway, we don't need to acquire it.

Reply:
I see! Thanks a lot for the analysis here, makes a lot of sense now.